File: bnbagent/utils/rate_limit.py
"""In-memory sliding-window rate limiter for public agent endpoints.

A public `/negotiate`-style endpoint signs negotiation hashes with the
provider's wallet on every accepted request. Without throttling, any caller
can drive arbitrary signing work and accumulate signed quotes; this limiter
caps the per-IP rate to bound that abuse without breaking marketplace
discovery.

The limiter is transport-agnostic: it raises :class:`RateLimitExceeded`,
and the serving layer (FastAPI route, MCP tool, A2A handler, ...) converts
that into its own protocol's rejection (e.g. HTTP 429).

Trade-offs (intentional, single-replica scope):
- In-memory state: counters are not shared across replicas. Multi-replica
  deployments effectively get N × the per-replica limit; that is acceptable
  while horizontal scaling itself raises the cost of an attack.
- Memory growth is hard-capped by ``max_keys``; least-recently-used keys
  are evicted once the cap is exceeded. This prevents an IPv6 address-cycling
  attack from growing the limiter's memory without bound.
"""

from __future__ import annotations

import time
from collections import OrderedDict, deque


class RateLimitExceeded(Exception):
    """Raised by :meth:`SlidingWindowLimiter.check` when a key's window is full.

    Transport-agnostic on purpose — the serving layer maps it to its own
    rejection (HTTP 429, MCP error, ...).
    """


class SlidingWindowLimiter:
    """Per-key sliding-window rate limiter with LRU key eviction.

    Allows up to ``max_requests`` events per ``window_seconds`` for any
    given key. Raises :class:`RateLimitExceeded` once the budget is exhausted.
    The number of tracked keys is hard-capped at ``max_keys``; once exceeded,
    the least-recently-used key is evicted to reclaim memory.
    """

    def __init__(self, max_requests: int, window_seconds: float, max_keys: int = 10_000) -> None:
        if max_requests <= 0:
            raise ValueError("max_requests must be > 0")
        if window_seconds <= 0:
            raise ValueError("window_seconds must be > 0")
        if max_keys <= 0:
            raise ValueError("max_keys must be > 0")
        self._max = max_requests
        self._window = window_seconds
        self._max_keys = max_keys
        self._buckets: OrderedDict[str, deque[float]] = OrderedDict()

    @property
    def max_requests(self) -> int:
        return self._max

    @property
    def window_seconds(self) -> float:
        return self._window

    @property
    def max_keys(self) -> int:
        return self._max_keys

    def check(self, key: str) -> None:
        """Record a hit for ``key`` or raise ``RateLimitExceeded`` if the window is full."""
        now = time.monotonic()
        if key in self._buckets:
            # Move to end to mark as most-recently used
            self._buckets.move_to_end(key)
            bucket = self._buckets[key]
        else:
            bucket = deque()
            self._buckets[key] = bucket
            # Evict least-recently-used key if cap is exceeded
            if len(self._buckets) > self._max_keys:
                self._buckets.popitem(last=False)
        cutoff = now - self._window
        while bucket and bucket[0] <= cutoff:
            bucket.popleft()
        if len(bucket) >= self._max:
            raise RateLimitExceeded("Too many requests")
        bucket.append(now)

File: typescript/src/utils/rateLimit.ts
/**
 * In-memory sliding-window rate limiter for public agent endpoints.
 *
 * A public `/negotiate`-style endpoint signs negotiation hashes with the
 * provider's wallet on every accepted request. Without throttling, any caller
 * can drive arbitrary signing work and accumulate signed quotes; this limiter
 * caps the per-IP rate to bound that abuse without breaking marketplace
 * discovery.
 *
 * The limiter is transport-agnostic: it throws {@link RateLimitExceeded},
 * and the serving layer (HTTP route, MCP tool, A2A handler, ...) converts
 * that into its own protocol's rejection (e.g. HTTP 429).
 *
 * Trade-offs (intentional, single-replica scope):
 * - In-memory state: counters are not shared across replicas. Multi-replica
 *   deployments effectively get N x the per-replica limit; that is acceptable
 *   while horizontal scaling itself raises the cost of an attack.
 * - Memory growth is hard-capped by `maxKeys`; least-recently-used keys
 *   are evicted once the cap is exceeded. This prevents an IPv6 address-cycling
 *   attack from growing the limiter's memory without bound.
 */

/**
 * Thrown by {@link SlidingWindowLimiter.check} when a key's window is full.
 *
 * Transport-agnostic on purpose — the serving layer maps it to its own
 * rejection (HTTP 429, MCP error, ...). Deliberately NOT a {@link
 * BNBAgentError}: this is a generic utility error outside the SDK's
 * protocol error hierarchy.
 */
export class RateLimitExceeded extends Error {
  constructor(message?: string) {
    super(message);
    this.name = "RateLimitExceeded";
  }
}

const DEFAULT_MAX_KEYS = 10_000;

/**
 * Per-key sliding-window rate limiter with LRU key eviction.
 *
 * Allows up to `maxRequests` events per `windowSeconds` for any given key.
 * Throws {@link RateLimitExceeded} once the budget is exhausted. The number
 * of tracked keys is hard-capped at `maxKeys`; once exceeded, the
 * least-recently-used key is evicted to reclaim memory.
 */
export class SlidingWindowLimiter {
  private readonly maxRequestsValue: number;
  private readonly windowSecondsValue: number;
  private readonly maxKeysValue: number;
  private readonly now: () => number;
  private readonly buckets = new Map<string, number[]>();

  constructor(
    maxRequests: number,
    windowSeconds: number,
    maxKeys: number = DEFAULT_MAX_KEYS,
    now: () => number = () => performance.now() / 1000,
  ) {
    if (maxRequests <= 0) {
      throw new Error("max_requests must be > 0");
    }
    if (windowSeconds <= 0) {
      throw new Error("window_seconds must be > 0");
    }
    if (maxKeys <= 0) {
      throw new Error("max_keys must be > 0");
    }
    this.maxRequestsValue = maxRequests;
    this.windowSecondsValue = windowSeconds;
    this.maxKeysValue = maxKeys;
    this.now = now;
  }

  get maxRequests(): number {
    return this.maxRequestsValue;
  }

  get windowSeconds(): number {
    return this.windowSecondsValue;
  }

  get maxKeys(): number {
    return this.maxKeysValue;
  }

  /** Record a hit for `key`, or throw {@link RateLimitExceeded} if its window is full. */
  check(key: string): void {
    const now = this.now();

    let bucket = this.buckets.get(key);
    if (bucket !== undefined) {
      // Touch: move to the end of the Map's insertion order to mark
      // as most-recently used.
      this.buckets.delete(key);
      this.buckets.set(key, bucket);
    } else {
      bucket = [];
      this.buckets.set(key, bucket);
      // Evict the least-recently-used key (the Map's first entry) if the
      // cap is exceeded.
      if (this.buckets.size > this.maxKeysValue) {
        const oldestKey = this.buckets.keys().next().value;
        if (oldestKey !== undefined) {
          this.buckets.delete(oldestKey);
        }
      }
    }

    const cutoff = now - this.windowSecondsValue;
    while (bucket.length > 0 && bucket[0] <= cutoff) {
      bucket.shift();
    }

    if (bucket.length >= this.maxRequestsValue) {
      throw new RateLimitExceeded("Too many requests");
    }
    bucket.push(now);
  }
}
